server/base/src/main/java/org/apache/accumulo/server/conf/codec/PropEncoding.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.server.conf.codec;
+
+import java.time.Instant;
+import java.util.Map;
+
+public interface PropEncoding {
+
+  /**
+   * Add a property. If the property already exists it is overwritten.
+   *
+   * @param key
+   *          the name of the property
+   * @param value
+   *          the value of the property.
+   */
+  void addProperty(String key, String value);
+
+  /**
+   * Add multiple properties. If a property already exists it is overwritten.
+   *
+   * @param properties
+   *          A map of key, value pairs.
+   */
+  void addProperties(Map<String,String> properties);
+
+  /**
+   * Get a store property or null if it does not exist.
+   *
+   * @param key
+   *          the name of the property.
+   * @return the property value.
+   */
+  String getProperty(String key);
+
+  /**
+   * Delete a property.
+   *
+   * @param key
+   *          the name of the property.
+   * @return the previous value if the property was present.
+   */
+  String removeProperty(String key);
+
```

Comment:
I will look into this - originally I was shooting to have the header state and the  properties encapsulated to ease management in the cache. If the cache only maintains a Map with the values then some other operations will need to track the header values separately.  The header data version is used to catch overlapping writes to ZooKeeper and should be maintained along with the property collection.

It may be better to separate the serialization (following your suggestion), but also keep the idea of a versioned collection of properties. Would need a better name though.

Reply:
Ah. I think I understand. This is looking a bit "MVC"-like:

1. Model - a versioned properties object (properties map + header info)
2. View - the serialization / encoding stuff (but more than a "view", because it's bidirectional)
3. Controller - the cache that holds a reference to an instance of the data, and triggers the read/write through the encoder

Whether or not the MVC pattern fits exactly, I do think there's some value in trying to keep the "worker" code separate from the data model, which is the core of what I was suggesting above, and I think still holds, even if you're working with a `VersionedProperties` object instead of a `Map<String,String>` object.

I actually like the name `VersionedProperties`, because you can easily have a `getVersion()` method on it that gives you something to compare in the cache. This version would be the "modification version", of course, not the "serialization format version", which could be entirely handled by the serializer/encoder code. The difference is that the "modification version" is relevant to determine whether the properties have changed, and the "serialization format version" has to do with backwards compatibility at the persistence layer.